File: Kalorie/Kalorie/Features/AddFoodSheet/AddFoodSheetConfigurator.swift
//
//  AddFoodSheetConfigurator.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation

struct AddFoodSheetConfigurator {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func createView(
        date: Date,
        mealTypes: [MealTypeDomain],
        onFoodSaved: @escaping () -> Void = {},
        withBarcodeScan: Bool = false
    ) -> AddFoodSheetView {
        let mealEditorConfigurator = MyCreatedMealEditorConfigurator(dataProvider: dataProvider, authProvider: authProvider)
        return AddFoodSheetView(
            viewModel: AddFoodSheetViewModel(
                searchFoodItems: SearchFoodItemsUseCase(dataProvider: dataProvider),
                submitFoodItem: SubmitFoodItemUseCase(dataProvider: dataProvider, authProvider: authProvider),
                fetchMySubmissions: FetchMySubmissionsUseCase(dataProvider: dataProvider, authProvider: authProvider),
                updateMySubmission: UpdateMySubmissionUseCase(dataProvider: dataProvider, authProvider: authProvider),
                deleteMySubmission: DeleteMySubmissionUseCase(dataProvider: dataProvider, authProvider: authProvider),
                searchFoodExternally: SearchFoodExternallyUseCase(),
                fetchFoodItemByBarcode: FetchFoodItemByBarcodeUseCase(dataProvider: dataProvider),
                fetchFoodByBarcodeExternally: FetchFoodByBarcodeExternallyUseCase(),
                fetchFavouriteFoods: FetchFavouriteFoodsUseCase(dataProvider: dataProvider, authProvider: authProvider),
                refreshFavouriteFood: RefreshFavouriteFoodUseCase(dataProvider: dataProvider, authProvider: authProvider),
                fetchMyCreatedMeals: FetchMyCreatedMealsUseCase(dataProvider: dataProvider, authProvider: authProvider),
                deleteMyCreatedMeal: DeleteMyCreatedMealUseCase(dataProvider: dataProvider, authProvider: authProvider),
                recognizeNutritionLabel: RecognizeNutritionLabelUseCase(),
                cameraAuthorizationProvider: CameraAuthorizationProvider(),
                onFoodSaved: onFoodSaved,
                isScannerVisible: withBarcodeScan
            ),
            makeFoodQuantityView: { [self] item, isFavourite, meal, onSaved, onFavouriteChanged, onMealUpdated in
                FoodQuantityView(
                    viewModel: FoodQuantityViewModel(
                        item: item,
                        saveFoodConsumed: SaveFoodConsumedUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        fetchMealTypes: FetchMealTypesUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        selectedDate: date,
                        mealTypes: mealTypes,
                        isFavourite: isFavourite,
                        addFavouriteFood: AddFavouriteFoodUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        removeFavouriteFood: RemoveFavouriteFoodUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        fetchFoodItemPersonalPortions: FetchFoodItemPersonalPortionsUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        saveFoodItemPersonalPortions: SaveFoodItemPersonalPortionsUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        fetchMyFoodItemReport: FetchMyFoodItemReportUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        submitFoodItemReport: SubmitFoodItemReportUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        meal: meal,
                        updateMyCreatedMeal: UpdateMyCreatedMealUseCase(dataProvider: dataProvider, authProvider: authProvider),
                        onSaved: onSaved,
                        onMealUpdated: onMealUpdated,
                        onFavouriteChanged: onFavouriteChanged,
                        quantity: meal != nil ? item.weight : (item.portions.isEmpty ? 100 : 1),
                        unit: meal != nil ? .grams : FoodQuantityViewModel.defaultUnit(for: item)
                    )
                )
            },
            makeMealEditorView: { onSaved in
                mealEditorConfigurator.createView(onSaved: onSaved, dismissesOnSave: false)
            }
        ) { meal, onSaved in
            mealEditorConfigurator.createView(existingMeal: meal, onSaved: onSaved)
        }
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/addfoodsheet/AddFoodSheetConfigurator.kt
package antoni.kalorie.features.addfoodsheet

import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.lifecycle.viewmodel.compose.viewModel
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.usecases.AddFavouriteFoodUseCase
import antoni.kalorie.core.usecases.FetchFavouriteFoodsUseCase
import antoni.kalorie.core.usecases.FetchFoodItemPersonalPortionsUseCase
import antoni.kalorie.core.usecases.FetchFoodByBarcodeExternallyUseCase
import antoni.kalorie.core.usecases.FetchFoodItemByBarcodeUseCase
import antoni.kalorie.core.usecases.FetchMealTypesUseCase
import antoni.kalorie.core.usecases.RefreshFavouriteFoodUseCase
import antoni.kalorie.core.usecases.RemoveFavouriteFoodUseCase
import antoni.kalorie.core.usecases.SaveFoodConsumedUseCase
import antoni.kalorie.core.usecases.SaveFoodItemPersonalPortionsUseCase
import antoni.kalorie.core.usecases.SearchFoodExternallyUseCase
import antoni.kalorie.core.usecases.SearchFoodItemsUseCase
import antoni.kalorie.features.foodquantity.FoodQuantityView
import antoni.kalorie.features.foodquantity.FoodQuantityViewModel
import java.time.Instant
import java.util.UUID

class AddFoodSheetConfigurator(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) {

    // MARK: - Functions

    @Composable
    fun createView(date: Instant, mealTypes: List<MealTypeDomain>, onDismiss: () -> Unit, onFoodSaved: () -> Unit = {}) {
        val instanceKey = remember { UUID.randomUUID().toString() }
        val viewModel = viewModel(key = instanceKey) {
            AddFoodSheetViewModel(
                searchFoodItems = SearchFoodItemsUseCase(dataProvider),
                searchFoodExternally = SearchFoodExternallyUseCase(),
                fetchFoodItemByBarcode = FetchFoodItemByBarcodeUseCase(dataProvider),
                fetchFoodByBarcodeExternally = FetchFoodByBarcodeExternallyUseCase(),
                fetchFavouriteFoods = FetchFavouriteFoodsUseCase(dataProvider, authProvider),
                refreshFavouriteFood = RefreshFavouriteFoodUseCase(dataProvider, authProvider),
                onFoodSaved = onFoodSaved,
            )
        }
        AddFoodSheetView(
            viewModel = viewModel,
            onDismiss = onDismiss,
            makeFoodQuantityView = { item, isFavourite, onSaved, onFavouriteChanged, onBack ->
                val quantityViewModel = viewModel {
                    FoodQuantityViewModel(
                        item = item,
                        saveFoodConsumed = SaveFoodConsumedUseCase(dataProvider, authProvider),
                        fetchMealTypes = FetchMealTypesUseCase(dataProvider, authProvider),
                        selectedDate = date,
                        mealTypes = mealTypes,
                        isFavourite = isFavourite,
                        addFavouriteFood = AddFavouriteFoodUseCase(dataProvider, authProvider),
                        removeFavouriteFood = RemoveFavouriteFoodUseCase(dataProvider, authProvider),
                        fetchFoodItemPersonalPortions = FetchFoodItemPersonalPortionsUseCase(dataProvider, authProvider),
                        saveFoodItemPersonalPortions = SaveFoodItemPersonalPortionsUseCase(dataProvider, authProvider),
                        onSaved = onSaved,
                        onFavouriteChanged = onFavouriteChanged,
                        quantity = if (item.portions.isEmpty()) 100.0 else 1.0,
                        unit = FoodQuantityViewModel.defaultUnit(item),
                    )
                }
                FoodQuantityView(viewModel = quantityViewModel, onBack = onBack)
            },
        )
    }
}
